runtime/memprof.c

```diff
@@ -654,33 +730,54 @@ void caml_memprof_track_young(uintnat wosize, int from_caml)
   }
 
   /* This condition happens either in the case of an exception or if
-     the callback returned [None]. If these cases happen frequently,
-     then we need to call [flush_deleted] somewhere to prevent a
-     leak. */
-  if (t_idx == Invalid_index)
+     one of the callbacks returned [None]. If these cases happen
+     frequently, then we need to call [flush_deleted] somewhere to
+     prevent a leak. */
+  if (has_delete)
     flush_deleted();
 
-  caml_raise_if_exception(res);
+  if (Is_exception_result(res)) {
+    for (i = 0; i < allocs_sampled; i++)
+      if (idx_tab[i] != Invalid_index) {
+        struct tracked* t = &trackst.entries[idx_tab[i]];
+        /* The allocations are cancelled because of the exception,
+           but this callback has already been called. We simulate a
+           deallocation. */
```

Comment:
If I understand correctly, the purpose is to allow a deallocation callback to be run later on. Is there a strict technical reason for this, or is it a design choice proposed for discussion?

Reply:
The typical callbacks for Memprof will add sampled value to a global table, and remove them on deallocation. (At any point you can show the table to the user, which contains a representative view of the live heap.) If you call allocation callbacks for certain allocation points and never call the deallocation callback, you have a leak in your global table. So it's a good guarantee to give that allocation-deallocation callbacks are well-bracketed while the program is running (that is: except possibly at program-termination time).

Reply:
I am convinced by your rationale in the normal conditions. But in the event of an interrupt, as a general rule, the only guarantees that the runtime ought to bring is resource-safety: anything that witnesses an interrupt is about to get destroyed anyway.

To me an exception arising from a Memprof callback is an exceptional condition where normal guarantees of Memprof operation are no longer satisfied. Indeed, an exception arising from an asynchronous callback is an interrupt (in SML parlance), the strongest form of exception, which should only caught at isolation boundaries if at all. By that point, the whole of Memprof instance already went bust probably or is about to, and the user has already marked their memprof data as invalid. (Unless the user has a very specific use-case and they know what they do, cf. allocation limits.) So while it is true that resource-safety must hold in the case of interrupts, this is for actual resources that are handled using dedicated and orthogonal features of the language such as unwind-protect wrappers. Therefore, to be resource-safe, one only needs to: 1) be resource-safe on the C side, and 2) make sure to properly propagate exceptions to destroy invalid state and call all release functions. It is true that resource-management in the presence of interrupts in not perfect currently, but this is not going to be fixed just for Memprof like this.

I suspect there is much in common with the other disucssion, would you like that we postpone this point and add it to the whole "meaning of exceptions arising from memprof" discussion which it would be useful to have at some point?

Reply:
I do buy your argument that callback-failure is an error situation not dissimilar to program termination. I discussed this again with Jacques-Henri and he points out that the current implementation is close-to-observationationally-equivalent to non-combined allocations; this is also a good property, so I'm happy to have the code as-is (even though your argument is reasonable that, had this feature not been included in the first place, we would not need to ask for it).

Reply:
It is true that my argument also means that whatever you do, it does not really matter...

Reply:
I agree with @gadmm's analysis here, but I'm happy to keep the current behaviour (making Comballoc transparent is nice).